src/unix/os390-syscalls.c

```diff
@@ -0,0 +1,271 @@
+struct _epoll_list* _global_epoll_list[MAX_EPOLL_INSTANCES];
+
+int scandir(const char *maindir, struct dirent ***namelist,
+            int (*filter)(const struct dirent *),
+            int (*compar)(const struct dirent **,
+            const struct dirent **)) {
+  struct dirent **nl = NULL;
+  struct dirent *dirent;
+  size_t count = 0;
+  size_t allocated = 0;
+  DIR *mdir;
+
+  mdir = opendir(maindir);
+  if (!mdir)
+    return -1;
+
+  while (1) {
+    dirent = readdir(mdir);
+    if (!dirent)
+      break;
+    if (!filter || filter(dirent)) {
+      struct dirent *copy;
+      copy = uv__malloc(sizeof(*copy));
+      if (!copy) {
+        while (count) {
+          dirent = nl[--count];
+          uv__free(dirent);
+        }
+        uv__free(nl);
+        closedir(mdir);
+        errno = ENOMEM;
+        return -1;
+      }
+      memcpy(copy, dirent, sizeof(*copy));
+
+      nl = uv__realloc(nl, count+1);
+      nl[count++] = copy;
+    }
+  }
+
+  qsort(nl, count, sizeof(struct dirent *),
+      (int (*)(const void *, const void *))compar);
+
+  closedir(mdir);
+
+  *namelist = nl;
+  return count;
+}
+
+static int isfdequal(const void* first, const void* second) {
+  const struct pollfd* a = (const struct pollfd*) first;
+  const struct pollfd* b = (const struct pollfd*) second;
+  return a->fd == b->fd ? 0 : 1;
+}
+
+int epoll_create1(int flags)
+{
+  struct _epoll_list* p = uv__malloc(sizeof(struct _epoll_list));
+
+  memset(p, 0, sizeof(struct _epoll_list));
+  int index = number_of_epolls++;
+  _global_epoll_list[index] = p;
+
+  if (uv_mutex_init(&p->lock)) {
+    errno = ENOLCK;
+    return -1;
+  }
+
+  p->size = 0;
+  return index; 
+}
+
+int epoll_ctl(int epfd, int op, int fd, struct epoll_event *event)
+{
+  struct _epoll_list *lst = _global_epoll_list[epfd];
+
+  if (op == EPOLL_CTL_DEL) {
+
+    struct pollfd* found;
+    struct pollfd pfd;
+
+    pfd.fd = fd;
+    uv_mutex_lock(&lst->lock);
+    found = lfind(&pfd, &lst->items[0], &lst->size,
+                  sizeof(struct pollfd), isfdequal);
+    if (found != NULL)
+      memcpy(found,
+             found+1,
+             (&lst->items[lst->size--] - found - 1) * sizeof(*found));
+    uv_mutex_unlock(&lst->lock);
+
+    if (found == NULL)
+      return ENOENT;
+
+  }
+  else if(op == EPOLL_CTL_ADD)
+  {
+    struct pollfd* found;
+    struct pollfd pfd;
+    size_t before;
+    size_t after;
+
+    if (lst->size == MAX_ITEMS_PER_EPOLL - 1)
+      return ENOMEM;
+
+    pfd.fd = fd;
+    pfd.events = event->events;
+    uv_mutex_lock(&lst->lock);
+    before = lst->size; 
+    found = lsearch(&pfd, &lst->items[0], &lst->size,
+                    sizeof(struct pollfd), isfdequal);
+    after = lst->size; 
+    uv_mutex_unlock(&lst->lock);
+
+    if (found != NULL && before == after) {
+      errno = EEXIST;
+      return -1;
+    }
+
+  }
+  else if(op == EPOLL_CTL_MOD)
+  {
+    int index;
+    struct pollfd* found;
+    struct pollfd pfd;
+
+    pfd.fd = fd;
```

Comment:
i see this construct (finding the right pollfd) used in a couple of places. can we factor it out?